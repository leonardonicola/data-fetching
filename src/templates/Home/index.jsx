import "./styles.css";
import Cards from "../../components/Cards";
import { useEffect, useState } from "react";
import fetchApi from "../../utils/fetch-api";
/* import loadPost from "./fetch/fetch-axios"; */

function App() {
  const [users, setUsers] = useState([]);
  const [counter, setCounter] = useState(4);
  const [search, setSearchName] = useState("");

  //fetching data via Axios

  /* 
  const handleFetchAxios = async () => {
    const { data } = await loadPost();
    setUsers(data);
  }; 

  useEffect(() => {
    handleFetchAxios();
  }, []); 
  */

  //fetching data via Fetch API

  useEffect(() => {
    const handleFetchApi = async () => {
      const data = await fetchApi();

      //if you want to get data doing SSR, just filter the data before setting it
      //this is going to give you much performance

      /*const filteredData = data.filter(el => el.id <= counter)
      setUsers(filteredData); */

      setUsers(data);
    };

    handleFetchApi();
  }, [counter]);

  return (
    <div className="app">
      <label>
        <h1>Pesquise: </h1>
        <input
          data-testid="search"
          type="text"
          value={search}
          onChange={(e) => setSearchName(e.target.value)}
        />
      </label>
      <div className="container" data-testid="posts">
        {(users.length === 0 && <h1>Sem posts!</h1>) || (!!search &&
          users
            .filter((el) => el.title.toLowerCase().includes(search.toLowerCase()))
            .map((el) => (
              <Cards title={el.title} body={el.body} id={el.id} />
            ))) ||
        (!!users &&
          users
            .filter((el) => el.id <= counter)
            .map((el) => (
              <Cards title={el.title} body={el.body} id={el.id} />
            ))) 
          }
      </div>
      {!!search || (
        <button onClick={() => setCounter((prevState) => prevState + 4)}>
          VER MAIS
        </button>
      )}
    </div>
  );
}

export default App;
