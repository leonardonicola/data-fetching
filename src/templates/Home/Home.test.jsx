import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Home from ".";
import { rest } from "msw";
import { setupServer } from "msw/node";

const handlers = [
  rest.get(
    "https://jsonplaceholder.typicode.com/posts",
    async (req, res, ctx) => {
      return res(
        ctx.json([
          {
            userId: 1,
            id: 1,
            title: "title 1",
            body: "body 1",
          },
          {
            userId: 2,
            id: 2,
            title: "title 2",
            body: "body 2",
          },
          {
            userId: 3,
            id: 3,
            title: "title 3",
            body: "body 3",
          },
          {
            userId: 4,
            id: 4,
            title: "title 4",
            body: "body 4",
          },
          {
            userId: 5,
            id: 5,
            title: "title 5",
            body: "body 5",
          },
        ])
      );
    }
  ),
];

const server = setupServer(...handlers);

describe("<Home/>", () => {
  beforeAll(() => server.listen());

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it("should have empty search name", () => {
    const { getByTestId } = render(<Home />);
    expect(getByTestId("search")).toBeEmptyDOMElement();
  });

  it("should have the input name equals title", () => {
    const { getByTestId } = render(<Home />);

    const input = getByTestId("search");
    const valorQualquer = "lorem";
    userEvent.type(input, valorQualquer);
    expect(valorQualquer).toEqual("lorem");
  });

  it("should render posts", async () => {
    render(<Home />);
    const noPosts = screen.getByText("Sem posts!");
    await waitForElementToBeRemoved(noPosts);
    expect(
      screen.getByRole("heading", { name: "title 1" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "title 2" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "title 3" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "title 4" })
    ).toBeInTheDocument();
  });

  it("should search for posts", async () => {
    render(<Home />);
    const noPosts = screen.getByText("Sem posts!");
    await waitForElementToBeRemoved(noPosts);
    const search = screen.getByTestId("search");
    userEvent.type(search, "title 1");

    expect(
      screen.getByRole("heading", { name: "title 1" })
    ).toBeInTheDocument();

    expect(
      screen.queryByRole("heading", { name: "title 2" })
    ).not.toBeInTheDocument();

    expect(
      screen.queryByRole("heading", { name: "title 3" })
    ).not.toBeInTheDocument();

    expect(
      screen.queryByRole("heading", { name: "title 4" })
    ).not.toBeInTheDocument();
  });

  it("shouldn't have value on input", async () => {
    render(<Home />);
    const noPosts = screen.getByText("Sem posts!");
    await waitForElementToBeRemoved(noPosts);
    const search = screen.getByTestId("search");
    expect(search.value).toBe("");
  });

  it("should load more posts", async () => {
    render(<Home />);
    const noPosts = screen.getByText("Sem posts!");
    await waitForElementToBeRemoved(noPosts);
    const button = screen.getByText("VER MAIS");
    userEvent.click(button);
    expect(
      screen.getByRole("heading", { name: "title 5" })
    ).toBeInTheDocument();
  });
});
