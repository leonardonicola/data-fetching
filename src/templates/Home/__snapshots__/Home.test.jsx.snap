// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> should match snapshot 1`] = `
<div
  class="app"
>
  <label>
    <h1>
      Pesquise: 
    </h1>
    <input
      data-testid="search"
      type="text"
      value=""
    />
  </label>
  <div
    class="container"
  />
  <button>
    VER MAIS
  </button>
</div>
`;
